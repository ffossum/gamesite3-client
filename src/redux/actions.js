/* @flow */
import type { RegistrationAction } from './modules/registration/registrationActions';
import type { LoginAction } from './modules/login/loginActions';
import type { ChatAction } from './modules/chat/chatActions';
import type { UserDataAction } from './modules/users/userDataActions';
import type { LobbyAction } from './modules/lobby/lobbyActions';

export type Action =
  | ChatAction
  | LoginAction
  | RegistrationAction
  | UserDataAction
  | LobbyAction
  | { type: '@@INIT' };
