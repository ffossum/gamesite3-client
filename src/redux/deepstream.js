/* @flow */
import deepstream from 'deepstream.io-client-js';

export function login() {
  return new Promise((resolve, reject) => {
    const client = deepstream('localhost:6020');
    client.login(success => {
      if (success) {
        resolve(client);
      }
      reject();
    });
  });
}
