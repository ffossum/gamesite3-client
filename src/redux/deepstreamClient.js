/* @flow */
import { Observable } from 'rxjs';

export default class DeepstreamClient {
  client: any;
  constructor(deepstream: any, ...args: any[]) {
    this.client = deepstream(...args);
  }
  login(...args: any[]): Promise<DeepstreamClient> {
    return new Promise((resolve, reject) => {
      this.client.login(...args, success => {
        success ? resolve(this) : reject();
      });
    });
  }
  subscribe(eventName: string): Observable<*> {
    return Observable.fromEventPattern(
      handler => this.client.event.subscribe(eventName, handler),
      handler => this.client.event.unsubscribe(eventName, handler),
    );
  }
  emit(eventName: string, data?: mixed): void {
    this.client.event.emit(eventName, data);
  }
  make(rpcName: string, data?: mixed) {
    return new Promise((resolve, reject) => {
      this.client.rpc.make(rpcName, data, (error, result) => {
        if (error) {
          reject(error);
        }
        resolve(result);
      });
    });
  }
}
