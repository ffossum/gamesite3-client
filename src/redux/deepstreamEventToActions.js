/* @flow */
import {
  playerJoined,
  playerLeft,
  gameCanceled,
} from './modules/games/gameRoomActions';
import { receiveMessage } from './modules/chat/chatActions';
import type { Action } from './actions';

type PlayerJoinedEvent = {
  t: 'player-joined',
  p: {
    gid: string,
    uid: string,
  },
};

type PlayerLeftEvent = {
  t: 'player-left',
  p: {
    gid: string,
    uid: string,
  },
};

type GameCanceledEvent = {
  t: 'game-canceled',
  p: {
    gid: string,
  },
};

type ChatMessageEvent = {
  t: 'chatmsg',
  uid: string,
  txt: string,
  ch: string,
};

type DeepstreamEvent =
  | PlayerJoinedEvent
  | PlayerLeftEvent
  | ChatMessageEvent
  | GameCanceledEvent;

export default function deepstreamEventToActions(
  event: DeepstreamEvent,
): Array<Action> {
  switch (event.t) {
    case 'player-joined': {
      const gameId = event.p.gid;
      const userId = event.p.uid;
      return [playerJoined(userId, gameId)];
    }
    case 'player-left': {
      const gameId = event.p.gid;
      const userId = event.p.uid;
      return [playerLeft(userId, gameId)];
    }
    case 'game-canceled': {
      const gameId = event.p.gid;
      return [gameCanceled(gameId)];
    }
    case 'chatmsg': {
      const time = new Date().toISOString();
      return [receiveMessage(event, time)];
    }
    default:
      return [];
  }
}
