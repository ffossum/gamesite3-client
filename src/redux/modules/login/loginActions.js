/* @flow */
export const LOGIN_REQUEST = 'login/request';
export const LOGIN_SUCCESS = 'login/success';
export const LOGIN_FAILURE = 'login/failure';

import type { SessionUser } from '../session/sessionReducer';

export type Login = {
  email: string,
  password: string,
};
export type LoginRequestAction = {
  type: typeof LOGIN_REQUEST,
  payload: Login,
};
export function loginRequest(login: Login): LoginRequestAction {
  return {
    type: LOGIN_REQUEST,
    payload: login,
  };
}

type LoginSuccessAction = {
  type: typeof LOGIN_SUCCESS,
  payload: SessionUser,
};
export function loginSuccess(user: SessionUser): LoginSuccessAction {
  return {
    type: LOGIN_SUCCESS,
    payload: user,
  };
}

type LoginFailureAction = {
  type: typeof LOGIN_FAILURE,
};
export function loginFailure(): LoginFailureAction {
  return {
    type: LOGIN_FAILURE,
  };
}

export type LoginAction =
  | LoginRequestAction
  | LoginSuccessAction
  | LoginFailureAction;
