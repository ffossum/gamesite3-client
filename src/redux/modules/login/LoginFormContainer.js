/* @flow */
import { connect } from 'react-redux';
import RegistrationForm from '../../../components/login/LoginForm';
import { loginRequest } from './loginActions';

import type { State } from '../root';

export default connect(
  (state: State) => ({
    loading: state.registration.loading,
  }),
  (dispatch: Dispatch<*>) => ({
    logIn(...args) {
      dispatch(loginRequest(...args));
    },
  })
)(RegistrationForm);
