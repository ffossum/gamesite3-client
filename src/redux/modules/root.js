/* @flow */
import { combineReducers } from 'redux';
import { combineEpics } from 'redux-observable';
import registration from './registration/registrationReducer';
import type { RegistrationState } from './registration/registrationReducer';
import registrationEpic from './registration/registrationEpic';
import session from './session/sessionReducer';
import type { SessionState } from './session/sessionReducer';
import sessionEpic from './session/sessionEpic';

import login from './login/loginReducer';
import type { LoginState } from './login/loginReducer';
import loginEpic from './login/loginEpic';

import chat from './chat/chatReducer';
import type { ChatState } from './chat/chatReducer';
import chatEpic from './chat/chatEpic';

import users from './users/usersReducer';
import userDataEpic from './users/userDataEpic';
import type { UsersState } from './users/usersReducer';

import lobby from './lobby/lobbyReducer';
import lobbyEpic from './lobby/lobbyEpic';
import type { LobbyState } from './lobby/lobbyReducer';

import games from './games/gamesReducer';
import type { GamesState } from './games/gamesReducer';
import gameRoomEpic from './games/gameRoomEpic';

export type State = {
  chat: ChatState,
  login: LoginState,
  registration: RegistrationState,
  session: SessionState,
  users: UsersState,
  lobby: LobbyState,
  games: GamesState,
};

export const rootReducer = combineReducers({
  chat,
  login,
  registration,
  session,
  users,
  lobby,
  games,
});

export const rootEpic = combineEpics(
  chatEpic,
  loginEpic,
  registrationEpic,
  sessionEpic,
  userDataEpic,
  lobbyEpic,
  gameRoomEpic,
);
