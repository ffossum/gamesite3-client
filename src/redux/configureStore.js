/* @flow */
import { createStore, applyMiddleware, compose } from 'redux';
import { createEpicMiddleware } from 'redux-observable';
import { rootReducer, rootEpic } from './modules/root';
import type { State } from './modules/root';

type Dependencies = {
  ajax: any,
  deepstreamClient: any,
  location: any,
  devToolsCompose: any,
};

export default function configureStore(
  preloadedState?: State,
  dependencies: Dependencies
) {
  const composeEnhancers =
    (DEVELOPMENT && dependencies.devToolsCompose) || compose;

  const epicMiddleware = createEpicMiddleware(rootEpic, {
    dependencies,
  });

  const storeEnhancer = composeEnhancers(applyMiddleware(epicMiddleware));

  if (preloadedState) {
    return createStore(rootReducer, preloadedState, storeEnhancer);
  }

  return createStore(rootReducer, storeEnhancer);
}
