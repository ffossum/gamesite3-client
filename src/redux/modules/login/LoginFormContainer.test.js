/* @flow */
/* eslint-env jest */
import React from 'react';
import LoginFormContainer, {
  mapStateToProps,
  mapDispatchToProps,
} from './LoginFormContainer';
import { loginRequest } from './loginActions';

import { createStore } from 'redux';
import { rootReducer } from '../root';
import { Provider } from 'react-redux';
import { mount } from 'enzyme';

describe('login form container', () => {
  test('is a renderable React component', () => {
    const store = createStore(rootReducer);
    const rendered = mount(
      <Provider store={store}>
        <LoginFormContainer />
      </Provider>,
    );

    expect(rendered).toMatchSnapshot();
  });

  describe('mapStateToProps', () => {
    test('selects loading state', () => {
      const store = createStore(rootReducer);
      expect(mapStateToProps(store.getState()).loading).toBe(false);

      store.dispatch(
        loginRequest({ email: 'asdf@asdf.com', password: 'asdfasdf' }),
      );
      expect(mapStateToProps(store.getState()).loading).toBe(true);
    });
  });

  describe('mapDispatchToProps', () => {
    test('logIn dispatches loginRequest action', () => {
      const dispatch = jest.fn();
      const props = mapDispatchToProps(dispatch);
      props.logIn({
        email: 'qwer@qwer.com',
        password: 'qwerqwer',
      });

      expect(dispatch).toHaveBeenCalledTimes(1);
      expect(dispatch).toHaveBeenCalledWith(
        loginRequest({
          email: 'qwer@qwer.com',
          password: 'qwerqwer',
        }),
      );
    });
  });
});
