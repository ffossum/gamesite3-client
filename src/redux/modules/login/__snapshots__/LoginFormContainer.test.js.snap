// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`login form container is a renderable React component 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <Connect(LoginForm)>
    <LoginForm
      loading={false}
      logIn={[Function]}
    >
      <form
        onSubmit={[Function]}
      >
        <div>
          <label
            htmlFor="email"
          >
            Email
            <input
              id="email"
              name="email"
              onChange={[Function]}
              readOnly={false}
              type="email"
            />
          </label>
        </div>
        <div>
          <label
            htmlFor="password"
          >
            Password
            <input
              id="password"
              name="password"
              onChange={[Function]}
              readOnly={false}
              type="password"
            />
          </label>
        </div>
        <button
          disabled={false}
          type="submit"
        >
          Log in
        </button>
      </form>
    </LoginForm>
  </Connect(LoginForm)>
</Provider>
`;
